### PR TITLE
spec: misc normative timezone fixes

### PR DESCRIPTION
**Pull request by ryzokuken**

### CODE REVIEW

File: spec/timezone.html [hunk 1]
```diff
@@ -269,8 +290,7 @@ <h1>Temporal.TimeZone.prototype.toJSON ( )</h1>
       </p>
       <emu-alg>
         1. Let _timeZone_ be the *this* value.
-        1. Let _toString_ be ? Get(_timeZone_, *"toString"*).
-        1. Return ? Call(_toString_, _timeZone_).
+        1. Return ? TimeZoneToString(_timeZone_).
```
Comment (Ms2ger) [inline spec/timezone.html:293]: Tests:

- [ ] `toString` is `undefined`
- [ ] `ToString` is called on the result

File: spec/timezone.html [hunk 2]
```diff
@@ -282,8 +302,7 @@ <h1>Abstract operations</h1>
       <h1>CreateTemporalTimeZoneFromStatic ( _constructor_, _identifier_ )</h1>
       <emu-alg>
         1. Assert: ! IsValidTimeZoneName(_identifier_) is *true*.
-        1. If ! IsConstructor(_constructor_) is *false*, then
-          1. Throw a *TypeError* exception.
+        1. If ! IsConstructor(_constructor_) is *false*, throw a *TypeError* exception.
```
Comment (Ms2ger) [inline spec/timezone.html:305]: I slightly prefer the two-line form, but happy to pick either consistently.

File: spec/timezone.html [hunk 3]
```diff
@@ -354,11 +370,10 @@ <h1>TimeZoneFrom ( _identifier_ )</h1>
       <h1>GetOffsetNanosecondsFor ( _timeZone_, _absolute_ )</h1>
       <emu-alg>
         1. Let _getOffsetNanosecondsFor_ be ? Get(_timeZone_, *"getOffsetNanosecondsFor"*).
+        1. If _getOffsetNanosecondsFor_ is *undefined*, set _getOffsetNanosecondsFor_ to %Temporal.TimeZone.prototype.getOffsetNanosecondsFor%.
```
Comment (Ms2ger) [inline spec/timezone.html:373]: Test for this in all callers

File: spec/timezone.html [hunk 4]
```diff
@@ -354,11 +370,10 @@ <h1>TimeZoneFrom ( _identifier_ )</h1>
...
         1. Let _offsetNanoseconds_ be ? Call(_getOffsetNanosecondsFor_, _timeZone_, « _absolute_ »).
-        1. If Type(_offsetNanoseconds_) is not Number, then
-          1. Throw a *TypeError* exception.
-        1. If _offsetNanoseconds_ is not an integer Number value between −86400 × 10<sup>9</sup> and 86400 × 10<sup>9</sup> exclusive, then
-          1. Throw a *RangeError* exception.
+        1. If Type(_offsetNanoseconds_) is not Number, throw a *TypeError* exception.
```
Comment (Ms2ger) [inline spec/timezone.html:375]: Btw, do we want this, or call `ToNumber`?

File: spec/timezone.html [hunk 5]
```diff
@@ -403,6 +415,7 @@ <h1>GetTemporalAbsoluteFor ( _timeZone_, _dateTime_, _disambiguation_ )</h1>
       <h1>GetPossibleAbsolutesFor ( _timeZone_, _dateTime_ )</h1>
       <emu-alg>
         1. Let _getPossibleAbsolutesFor_ be ? Get(_timeZone_, *"getPossibleAbsolutesFor"*).
+        1. If _getPossibleAbsolutesFor_ is *undefined*, set _getPossibleAbsolutesFor_ to %Temporal.TimeZone.prototype.getPossibleAbsolutesFor%.
```
Comment (Ms2ger) [inline spec/timezone.html:418]: Test

**Review by Ms2ger — CHANGES_REQUESTED** [review]
Consider splitting the non-normative changes into a separate commit, if not a separate PR.

**Comment by Ms2ger** [comment]
Oh, and please check if the polyfill needs matching changes.

File: spec/timezone.html (hunk 1 above)
Comment (ryzokuken) [reply spec/timezone.html:293]: Will add.

File: spec/timezone.html (hunk 2 above)
Comment (ryzokuken) [reply spec/timezone.html:305]: I'm personally fine with either, but I had heard that the single line form was to be preferred, thus this change.

File: spec/timezone.html (hunk 3 above)
Comment (ryzokuken) [reply spec/timezone.html:373]: Will do.

File: spec/timezone.html (hunk 4 above)
Comment (ryzokuken) [reply spec/timezone.html:375]: Sounds fine, to me.

**Comment by ryzokuken** [comment]
> Consider splitting the non-normative changes into a separate commit, if not a separate PR.

#833

File: polyfill/lib/ecmascript.mjs [hunk 6]
```diff
@@ -1496,6 +1496,21 @@ export const ES = ObjectAssign({}, ES2019, {
...
+    const offsetNanoseconds = ES.Call(getOffsetNanosecondsFor, timezone, [absolute]);
+    if (typeof offsetNanoseconds !== 'number') {
+      throw new TypeError('bad return from getOffsetNanosecondsFor');
+    }
+    if (!Number.isInteger(offsetNanoseconds) || Math.abs(offsetNanoseconds) > 86400e9) {
+      throw new RangeError('out-of-range return from getOffsetNanosecondsFor');
+    }
+    return offsetNanoseconds;
+  },
+
```
Comment (Ms2ger) [inline polyfill/lib/ecmascript.mjs]:
```suggestion
```

```
/home/travis/build/tc39/proposal-temporal/polyfill/lib/ecmascript.mjs
  1499:3  error  Duplicate key 'GetOffsetNanosecondsFor'  no-dupe-keys
```

File: polyfill/lib/timezone.mjs [hunk 7]
```diff
@@ -244,4 +243,5 @@ DefineIntrinsic('Temporal.TimeZone.prototype.getDateTimeFor', TimeZone.prototype
 DefineIntrinsic('Temporal.TimeZone.prototype.getAbsoluteFor', TimeZone.prototype.getAbsoluteFor);
 DefineIntrinsic('Temporal.TimeZone.prototype.getOffsetNanosecondsFor', TimeZone.prototype.getOffsetNanosecondsFor);
 DefineIntrinsic('Temporal.TimeZone.prototype.getOffsetStringFor', TimeZone.prototype.getOffsetStringFor);
+DefineIntrinsic('Temporal.TimeZone.prototype.getOffsetNanosecondsFor', TimeZone.prototype.getOffsetNanosecondsFor);
```
Comment (Ms2ger) [inline polyfill/lib/timezone.mjs:234]:
```suggestion
```

And we need a check that intrinsics are only registered once.

File: polyfill/lib/timezone.mjs [hunk 8]
```diff
@@ -53,9 +53,8 @@ export class TimeZone {
     return ES.TimeZoneToString(this);
   }
   getOffsetNanosecondsFor(absolute) {
-    if (!ES.IsTemporalTimeZone(this)) throw new TypeError('invalid receiver');
     if (!ES.IsTemporalAbsolute(absolute)) throw new TypeError('invalid Absolute object');
-    const id = GetSlot(this, TIMEZONE_ID);
+    const id = ES.TimeZoneToString(this);
```
Comment (Ms2ger) [inline polyfill/lib/timezone.mjs:58]: It's not clear to me that we have consensus on this change.

File: polyfill/test/TimeZone/prototype/getDateTimeFor/getoffsetnanosecondsfor-call.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2020 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline polyfill/test/TimeZone/prototype/getDateTimeFor/getoffsetnanosecondsfor-call.js:1]: Compare `polyfill/test/TimeZone/prototype/getDateTimeFor/custom-timezone.js`

File: polyfill/test/TimeZone/prototype/getOffsetStringFor/getoffsetnanosecondsfor-call.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2020 Igalia, S.L. All rights reserved.
```
Comment (Ms2ger) [inline polyfill/test/TimeZone/prototype/getOffsetStringFor/getoffsetnanosecondsfor-call.js:1]: Compare `polyfill/test/TimeZone/prototype/getOffsetStringFor/custom-timezone.js`

File: polyfill/lib/ecmascript.mjs [hunk 11]
```diff
@@ -1510,6 +1510,17 @@ export const ES = ObjectAssign({}, ES2019, {
...
+  GetPossibleAbsolutesFor: (timezone, datetime) => {
+    let getPossibleAbsolutesFor = timezone.getPossibleAbsolutesFor;
+    if (getPossibleAbsolutesFor === undefined) {
+      getPossibleAbsolutesFor = GetIntrinsic('%Temporal.TimeZone.prototype.getPossibleAbsolutesFor%');
+    }
+    const possibleAbsolutes = ES.Call(getPossibleAbsolutesFor, timezone, [datetime]);
+    for (const absolute of possibleAbsolutes) {
+      if (!ES.IsTemporalAbsolute(absolute)) throw new TypeError('bad return from getPossibleAbsolutesFor');
+    }
+    return possibleAbsolutes;
```
Comment (Ms2ger) [inline polyfill/lib/ecmascript.mjs:1522]: This needs to create a new `Array`. The following value will cause issues:

```js
{
  [Symbol.iterator]() { return [][Symbol.iterator](); },
  get length() { throw new Test262Error("Should not get 'length' property"); },
}
```

File: polyfill/lib/timezone.mjs [hunk 12]
```diff
@@ -92,28 +92,17 @@ export class TimeZone {
     const disambiguation = ES.ToTimeZoneTemporalDisambiguation(options);
 
     const Absolute = GetIntrinsic('%Temporal.Absolute%');
-    const possibleAbsolutes = this.getPossibleAbsolutesFor(dateTime);
-    if (!Array.isArray(possibleAbsolutes)) {
-      throw new TypeError('bad return from getPossibleAbsolutesFor');
-    }
+    const possibleAbsolutes = ES.GetPossibleAbsolutesFor(this, dateTime);
```
Comment (Ms2ger) [inline polyfill/lib/timezone.mjs:95]: Note that there's another `getPossibleAbsolutesFor` call below.

File: polyfill/lib/timezone.mjs [hunk 13]
```diff
@@ -156,10 +145,9 @@ export class TimeZone {
     }
   }
   getPossibleAbsolutesFor(dateTime) {
-    if (!ES.IsTemporalTimeZone(this)) throw new TypeError('invalid receiver');
     if (!ES.IsTemporalDateTime(dateTime)) throw new TypeError('invalid DateTime object');
     const Absolute = GetIntrinsic('%Temporal.Absolute%');
-    const id = GetSlot(this, TIMEZONE_ID);
+    const id = ES.TimeZoneToString(this);
```
Comment (Ms2ger) [inline polyfill/lib/timezone.mjs:150]: Not sure we have consensus for this either.

File: polyfill/test/Absolute/constructor/from/getpossibleabsolutesfor-call.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
+// Copyright (C) 2020 Igalia, S.L. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal.absolute.from
+includes: [compareArray.js]
+---*/
+
+const actual = [];
+const expected = ['get Temporal.TimeZone.from', 'call Temporal.TimeZone.from', 'get timeZone.getPossibleAbsolutesFor', 'call timeZone.getPossibleAbsolutesFor'];
```
Comment (Ms2ger) [inline polyfill/test/Absolute/constructor/from/getpossibleabsolutesfor-call.js:10]: This is getting hard to read and edit. What's wrong with one event per line?

File: polyfill/test/Absolute/constructor/from/getpossibleabsolutesfor-call.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+const abs = Temporal.now.absolute();
```
Comment (Ms2ger) [inline polyfill/test/Absolute/constructor/from/getpossibleabsolutesfor-call.js:12]: I would generally avoid using `now` in tests that don't need it.

**Review by Ms2ger — CHANGES_REQUESTED** [review]

**Comment by ryzokuken** [comment]
Dropping the commits for `getOffsetNanosecondsFor` and `getPossibleAbsolutesFor` (since there isn't consensus for changing them) and keeping the `toJSON`, it seems like everything has been done by @Ms2ger already. What remains is the addition of a single test (of testing `toJSON` on a custom timezone without `toString`) that I'll update this PR with, but feel free to just close this.

**Review by ptomato — APPROVED** [review]
:+1:, this makes sense to me.

**Pull request merged**